File: unit-tests/log/test-mixed-file-callback.cpp
// License: Apache 2.0. See LICENSE file in root directory.
// Copyright(c) 2020 Intel Corporation. All Rights Reserved.

//#cmake: static!

//#cmake:add-file log-common.h
#include "log-common.h"


TEST_CASE( "Mixed file & callback logging", "[log]" ) {

    char filename[L_tmpnam];
    tmpnam( filename );

    TRACE( "Filename logging to: " << filename );
    REQUIRE_NOTHROW( rs2::log_to_file( RS2_LOG_SEVERITY_ERROR, filename ));
    size_t n_callbacks = 0;
    REQUIRE_NOTHROW( rs2::log_to_callback( RS2_LOG_SEVERITY_ALL,
        [&]( rs2_log_severity severity, rs2::log_message const& msg )
        {
            ++n_callbacks;
            TRACE( severity << ' ' << msg.filename() << '+' << msg.line_number() << ": " << msg.raw() );
        } ));

    // Following should log to both
    log_all();

    el::Loggers::flushAll();   // requires static!
    REQUIRE( count_lines( filename ) == 1 );

    REQUIRE( n_callbacks == 4 );
}

File: unit-tests/log/test-mixed-file-callback.py
# License: Apache 2.0. See LICENSE file in root directory.
# Copyright(c) 2022 Intel Corporation. All Rights Reserved.

from rspy import log, test
import pyrealsense2 as rs
import common
import tempfile, os.path


#############################################################################################
#
test.start( f"Mixed file & callback logging" )

try:
    filename = tempfile.mktemp()
    log.d( f'Filename logging to: {filename}' )
    rs.log_to_file( rs.log_severity.error, filename )
    rs.log_to_callback( rs.log_severity.debug, common.message_counter )

    common.log_all()

    rs.reset_logger()  # Should flush!
    #el::Loggers::flushAll();   // requires static!

    test.check_equal( common.count_lines( filename ), 1 )
    test.check_equal( common.n_messages, 4 )
except:
    test.unexpected_exception()

test.finish()
#
#############################################################################################
test.print_results_and_exit()
